kframework/article.md

```diff
@@ -0,0 +1,169 @@
+
+```
+COST CENTRE  MODULE                    %time  %alloc
+function     Module.hs:(13,1)-(14,12)  3.8    1.3
+```
+
+The function `function`, defined in the module `Module` from line 13 column 1 to line 14 column 12, is responsible for 3.8% of the time cost and 1.3% of the allocation cost of the program.
+
+However, recall that by making certain functions into cost centres, their inlining is prevented. In a non-profiled executable they could be optimized away; therefore, the correct way to read the entry above is this:
+
+* **Unless** optimized away, `Module.function` **would** take 3.8% of time and occupy 1.3% of space.
+
+## Inlining
+
+Even without the help of the profiling instrumentation, we are still able to visually detect some parts of the code that can hinder GHC's abilities for optimization. Consider the following expression:
+
+```hs
+do a <- m
+   b <- n
+   f a b
+```
+
+Suppose the monad we work in is `State s`. Then the `>>=`s produced by the `do`-notation could be inlined and turned into `case`-expressions:
+
+```hs
+State $ \s ->
+  case runState m s of (a, s') ->
+  case runState n s' of (b, s'') ->
+  runState (f a b) s''
+```
+
+If, however, we work in some abstract monad `m` satisfying the `MonadState s m` constraint, then the `>>=`s are opaque to the compiler: it has to follow a pointer in the class dictionary and jump to an unknown function, which is bad for branch prediction.
+
+To find out if the functions in question are getting inlined, we can insert the following line to our module:
+
+```hs
+{-# OPTIONS -ddump-simpl -ddump-to-file #-}
+```
+
+1. `-ddump-to-file` causes the output from each of flags starting with `-ddump`, to be dumped to a file or files.
+2. `-ddump-simpl` dumps simplifier output.
+
+After building our module, GHC will produce a `.dump-simpl` file containing its internal representation of the program after most optimizations have had the chance to take place.
+
+Suppose we built the module where the `do`-notation snippet is defined, and we worked in an abstract monad `m`. When we search for `>>=` in our generated `.dump-simpl` file, we will stumble upon this snippet:
+
+```
+>>=
+  @ m_aM5
+  $dMonad_aMr
+  @ a_aM6
+  @ c_aM8
+  eta1_B2
+  (\ (a1_anu :: a_aM6) ->
+     >>=
+       @ m_aM5
+       $dMonad_aMr
+       @ b_aM7
+       @ c_aM8
+       eta2_B1
+       (\ (b1_anv :: b_aM7) -> eta_B3 a1_anu b1_anv))
+```
+
+The call to `>>=` is **not** inlined, and the reason is because it is instantiated at an abstract monad `m_aM5`, and GHC does not now which instance to use. Instead, it has to pass the implementation in a dictionary `$dMonad_aMr`.
+
+## Findings
+
+Here is a list of relevant terminology, datatypes, and typeclasses:
```

Comment:
It is odd to have a section called Findings that starts listing terminology at the start of it. 😅

Reply:
Right, I guess I need to refactor this part a little.
This list of terminology, datatypes, and typeclasses (let's call it _list1_) is relevant only to the _Specialization_ subsection.  The _Cache hit ratio_ and the _String interning_ subsections also contain their respective terminologies at the start, which is inconsistent with how _list1_ is placed (**before** the _Specialization_ subsection).

Which option is more logical?
1. To place _list1_ after the _Specialization_ header?
2. To place all definitions from **all** the aforementioned subsection to another standalone subsection (e.g. _Relevant Terminology, Datatypes, and Typeclasses_)?

P.S. I noticed another inconsistency in my draft: the differing capitalization rules in _Cache hit ratio_ and _String Interning_. Which capitalization style is preferred?

Reply:
Sentence case, e.g., Cache hit ratio for headings

Reply:
It's better to have the definition as close to the usage point as possible. So definitely not a big vocabulary as a subsection. 

The way you have it at the start of the each section is fine, I personally prefer for it to be weaved in the text more.